jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/JettyAppender.java

```diff
@@ -0,0 +1,24 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.logging;
+
+public interface JettyAppender
+{
+    void emit(JettyLoggingEvent event);
```

Comment:
I don't particularly like this because it is designed to _force_ allocation, and we can avoid it.
This can just be `emit(JettyLogger, Level, String, Object...)` as all the rest can be computed by the implementation.

Reply:
This is in preparation for the slf4j 2.0.0 behaviors.
Which will use the `org.slf4j.event.LoggingEvent` to handle the differences between normal loggers and location aware loggers, along with the markers and context key value pairs.

Reply:
I see this as just an implementation detail. 
Whether we will be forced by an implementation to create an event it's orthogonal to the design of this interface.
Events are not part of the SLF4J API so we don't need them.

Reply:
We can always have an Appender that makes events if they are needed by an integration

Reply:
The new interface would look like this ...

``` java
    void emit(Level level, long timestamp, String threadName, String name, String condensedName, String message, Throwable throwable, Object[] argumentArray, 
              boolean hideStacks, Marker marker);
```

And when slf4j 2.0.0 hits (currently in alpha, no public log4j 2 support yet) the API will look like this ...

``` java
    void emit(Level level, long timestamp, String threadName, String name, String condensedName, String message, Throwable throwable, Object[] argumentArray, 
              boolean hideStacks, List<Marker> markers, List<KeyValuePair> keyValueContextPairs);
```

Reply:
> We can always have an Appender that makes events if they are needed by an integration

Slf4j will give us `org.slf4j.event.LoggingEvent` instances from the API itself in version 2.0.0+
The idea of an Appender to make events isn't really a thing, we receive them, not the other way around.

Reply:
This is now feeling a lot more like we are using slf4j implementation rather than their interface, making me much less interested in this idea.
I just don't see why the implementation of `LOG.debug("foo");` should end up in being much more than a `System.err.printf` under the hood.   We don't want our default implementation to be a stream of slf4j event objects.

I don't understand your point about us not generating events.  The proposed code for debug is:
```java
     @Override
    public void debug(String msg)
    {
        if (isDebugEnabled())
        {
            getAppender().emit(asEvent(Level.DEBUG, msg));
        }
    }
```

But there is no reason that it cannot be:
```java
     @Override
    public void debug(String msg)
    {
        if (isDebugEnabled())
        {
            getAppender().emit(Level.DEBUG, msg);
        }
    }
```

and if events are needed, then emit can create them.... but in fact I can't even see why we need the abstraction of an appender? Shouldn't this code just be:


But there is no reason that it cannot be:
```java
     @Override
    public void debug(String msg)
    {
        if (isDebugEnabled())
        {
            emit(Level.DEBUG, msg);
        }
    }
```

If we want to do something more complex, then the whole implementation of debug is replaced with another jar.    We don't want to reinvent log interception/filtering/appending etc. etc.  we just want a simple StdErr implementation that is statically initialised and can be replaced with a third party implementation if anything sophisticated is desired.

Reply:
Splitting up the Logger from what actually produces the log (Appender) makes things easier to maintain and test, and as a bonus it uses less memory (as each logger instance is no longer doing the entire job of emitting the log).

I'll give it a go at making our Appender have multiple entry points.

And with slf4j 2.0.0  ...
We have Markers and KeyValuePairs as well to make a decision on. (do we care to support them or not).
Markers are supported by logback only (currently).
KeyValuePairs (the end result of MDC use) are supported by log4j1 impl, logback impl, and jdk/jul impls currently. (log4j2 impl in git shows signs of supporting them too).

Reply:
A split API for Appender would now look like ...

``` java
    void emit(JettyLogger logger, Level level, long timestamp, String threadName, String message);

    void emit(JettyLogger logger, Level level, long timestamp, String threadName,
              Throwable throwable, String message);

    void emit(JettyLogger logger, Level level, long timestamp, String threadName, String message,
              Object[] argumentArray);

    void emit(JettyLogger logger, Level level, long timestamp, String threadName, String name,
              Throwable throwable, String message, Object[] argumentArray);
```

Where each parameter is ...
* `JettyLogger logger` is used for the logger name (full or condensed based on appender config), and the hideStacks configuration for that logger.
* `Level level` is for the level of the event itself.
* `long timestamp` the timestamp at logging event time (either from the raw `Logger` methods like `.info()` and `.debug`, but also from `org.slf4j.event.LoggingEvent` data given to us)
* `String threadName` the threadName at logging event time (either from the raw `Logger` methods like `.info()` and `.debug`, but also from `org.slf4j.event.LoggingEvent` data given to us)
* `String message` the log message (or log message format)
* `Throwable throwable` the throwable given. (presented as a whole stacktrace or not, based on logger's hideStacks configuration)
* `Object[] argumentArray` the array of arguments given to the log.

Reply:
There is no need for all those parameters. The 4 I gave in the initial comment are sufficient.

Reply:
> This is now feeling a lot more like we are using slf4j implementation rather than their interface, making me much less interested in this idea.

The entirety of Jetty uses only `org.slf4j.Logger` and `org.slf4j.LoggerFactory`, we are using their interface.
This `jetty-slf4j-impl` is our implementation of those, and are completely optional.
In fact, only the `StacklessLogging` class is used directly by Jetty test cases, the rest is not used directly by Jetty.

**Scenario 1**

The use of slf4j also allows us to (for example) capture other loggers and route them to any slf4j implementation we want.

Jetty uses the `org.slf4j.Logger` API to create it's logging events.
We have the following jars on the classpath.
* `slf4j-api.jar`
* `jetty-slf4j-impl.jar` (this is the slf4j impl in use)
* `jcl-over-slf4j.jar` (bridge api for commons-logging)
* `jul-to-slf4j.jar` (bridge api for java.util.logging, JSP and mail jars use this)
* `log4j-to-slf4j.jar` (bridge api for log4j 1.x and log4j 2.x)

(For slf4j 1.x there are also bridge API jars for apache juli, and osgi logging, haven't checked for slf4j 2.x)

We can receive events from any of the logging APIs now.
Our `org.eclipse.jetty.logging.JettyLogger` can receive those events as ...
* Calls to equivalent `.info()` or `.debug()` style methods
* A call to `public void log(org.slf4j.event.LoggingEvent event)` (depends on bridge API - this is the `SubstituteLogger` internal slf4j API).
* A call to `public void log(Marker marker, String callerFQCN, int level, String msg, Object[] arguments, Throwable t)` (depends on bridge API and if source/location features - this is the `LocationAwareLogger` internal slf4j API)
The existence of the `org.slf4j.event.LoggingEvent` is to handle loggers that process events in a non-direct way:
* lost threadname information (like java.util.logging)
* timestamp skews (like osgi logging)
With this PR, the `jetty-slf4j-impl` can handle all of those events, and can now be written to `System.err` via the `StdErrAppender` (not supported in Jetty 9.4.x or current 10.x HEAD) and we don't have to care about where it came from, just implement the interface `org.slf4j.Logger` (with `LocationAwareLogger` and `SubstituteLogger` extensions)

**Scenario 2**

Jetty uses the `org.slf4j.Logger` API to create it's logging events.
The user has chosen to use it's own slf4j impl (lets call it `corp-slf4j-impl.jar`)
Jetty participates with logging level / name / MDC / location without extra work or effort.
All of the classes in this PR are unused and are not present.

Reply:
> There is no need for all those parameters. The 4 I gave in the initial comment are sufficient.

No, they are not.
Especially when using slf4j bridge APIs for capturing other loggers into our `jetty-slf4j-impl`

Reply:
The entire StdErrAppender API interface has been redone.
No more JettyLoggingEvent.